### ISSUE TITLE
Add makeIvyXml

### ISSUE THREAD

**Issue by mr-git** [sbt/sbt#3419]
We have not set publishing configuration, as we generate deliveries only, publishing in repo gets done by other means. 

In 0.13.16 calling `deliver` generated `ivy.xml` file only, but in 1.0.0 it generates `ivy.xml` and tries to publish too, which fails.

```
error] java.lang.RuntimeException: Repository for publishing is not specified.
[error] 	at scala.sys.package$.error(package.scala:27)
[error] 	at sbt.Classpaths$.$anonfun$getPublishTo$1(Defaults.scala:2396)
[error] 	at scala.Option.getOrElse(Option.scala:121)
[error] 	at sbt.Classpaths$.getPublishTo(Defaults.scala:2396)
[error] 	at sbt.Classpaths$.$anonfun$ivyBaseSettings$46(Defaults.scala:1890)
[error] 	at scala.Function1.$anonfun$compose$1(Function1.scala:44)
[error] 	at sbt.internal.util.$tilde$greater.$anonfun$$u2219$1(TypeFunctions.scala:42)
[error] 	at sbt.std.Transform$$anon$4.work(System.scala:64)
[error] 	at sbt.Execute.$anonfun$submit$2(Execute.scala:257)
[error] 	at sbt.internal.util.ErrorHandling$.wideConvert(ErrorHandling.scala:17)
[error] 	at sbt.Execute.work(Execute.scala:266)
[error] 	at sbt.Execute.$anonfun$submit$1(Execute.scala:257)
[error] 	at sbt.ConcurrentRestrictions$$anon$4.$anonfun$submitValid$1(ConcurrentRestrictions.scala:167)
[error] 	at sbt.CompletionService$$anon$2.call(CompletionService.scala:32)
[error] 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
[error] 	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
[error] 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
[error] 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
[error] 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
[error] 	at java.lang.Thread.run(Thread.java:745)
[error] (*:publishConfiguration) Repository for publishing is not specified.
```

Is it expected behaviour?

**Comment by mr-git** [comment]
`deliverLocal` seems to generate `ivy.xml` without attempting to publish it, but the content is abit different, than in one generated by 0.13.16 + `deliver`

**Comment by eed3si9n** [comment]
`deliver` blowing up is not an expected behavior, but there's a know issue (https://github.com/sbt/sbt/issues/2088#issuecomment-246208872) around the ordering of publishing ivy.xml, so I was thinking about eventually decommissioning the task, and fold it into `publish`.

**Comment by mr-git** [comment]
hm.. I do not want to `publish` as it involves uploading to some repo. I want to generate the `ivy.xml` file only. In my head `makePom` and `deliver` (which I would name `makeIvyXml`) have same objective: generate artefact descriptor for Maven or Ivy respectively.

**Comment by eed3si9n** [comment]
If I understand correctly, `deliver` task (which originally comes from Ivy) not only generates the `ivy.xml` but it actually send the file to the repository. Thus, it is named deliver. This is something I don't want to happen, because I won't have the control over the timing. See [deliver](http://ant.apache.org/ivy/history/2.3.0/use/deliver.html) for more detail.

We can try to provide `makeIvyXml` if Ivy lets us do it.

**Comment by mr-git** [comment]
So, does this means that Ivy changed its functionality between SBT 0.13.16 and 1.0.0 versions? Because in 0.13.16 `deliver` was only generating Ivy xml.

https://github.com/sbt/sbt/blob/6ea9dab99ec19f35f5635122e26d1fd3b06bba4a/main/src/main/scala/sbt/Keys.scala#L371 - here it says that is only generates, but doesn't mention that it wants to publish too.

**Comment by mr-git** [comment]
@eed3si9n, it is very interesting to read `deliver and publish` section in [deliver](http://ant.apache.org/ivy/history/2.3.0/use/deliver.html) :)

> So, when shall the deliver task be called explictly? When you actually need to separate what is performed by the deliver task (see above), from what is performed by the publish task, i.e. upload a module to a repository.

**Comment by eed3si9n** [comment]
> So, does this means that Ivy changed its functionality between SBT 0.13.16 and 1.0.0 versions? 

Both sbt 1.0 and 0.13.16 use Ivy.

Deliver doc http://ant.apache.org/ivy/history/2.3.0/use/deliver.html says:

> If a deliver target is set, then it is called (via an antcall) for each dependency which has not a sufficient status compared to the deliver status set for this task. This means that if you deliver an integration revision, no recursive delivery will be done.

It depends on the status. And here is how we set the status:

```scala
    deliverLocalConfiguration := deliverConfig(crossTarget.value, status = if (isSnapshot.value) "integration" else "release", logging = ivyLoggingLevel.value),
    deliverConfiguration := deliverLocalConfiguration.value,
```

**Comment by mr-git** [comment]
`isSnapshot` is `false` for both SBT versions, as expected.

```
    deliverLocalConfiguration := deliverConfig(crossTarget.value, status = if (isSnapshot.value) "integration" else "release", logging = ivyLoggingLevel.value),
    deliverConfiguration := deliverLocalConfiguration.value,
```

Strange, the above code is from SBT 0.13.16 (`Defaults.scala:1284..1285`)... but there are no such lines in 1.0.0 code... Am I looking in wrong place?

https://github.com/sbt/sbt/search?utf8=%E2%9C%93&q=deliverConfiguration&type= - empty
https://github.com/sbt/sbt/search?utf8=%E2%9C%93&q=deliverLocalConfiguration&type= - empty

**Comment by eed3si9n** [comment]
It's folded in here: https://github.com/sbt/sbt/blob/v1.0.0/main/src/main/scala/sbt/Defaults.scala#L1882-L1894

**Comment by mr-git** [comment]
https://github.com/sbt/sbt/commit/347914191a8af2f2ad53abbb120fb4144122380b - here the keys `deliverConfiguration` and `deliverLocalConfiguration` were removed and in Defaults.scala they were replaced with `publishConfiguration` - very nice, indeed...:
```
 -    deliver := deliverTask(deliverConfiguration).value,
 -    deliverLocal := deliverTask(deliverLocalConfiguration).value,
 +    deliver := deliverTask(publishConfiguration).value,
 +    deliverLocal := deliverTask(publishLocalConfiguration).value,
```

Now, could you please provide the hint on how to disable publishing or replace the publishing destination to some fake one in `target`?

**Comment by mr-git** [comment]
Ivy separates `deliver` and `publish` tasks, but why SBT decides to merge them both in `publish`?

**Comment by eed3si9n** [comment]
It's because we want to eventually control the precise timing on when deliver is called.
See https://github.com/sbt/sbt/issues/2088#issuecomment-246208872

**Comment by mr-git** [comment]
Ok, found workaround... `publishTo := Some(Resolver.file("file",  file("hack-sbt")))`
sadly, `publishTo := None` produces same exception: `Repository for publishing is not specified.`

But produced `ivy.xml` file contains a lot of lines that are meant for local delivery, not for public one.

Diff includes lines like:
```    <conf extends="compile,optional,provided" name="compile-internal" visibility="public" description="">

    <dependency org="org.scala-lang" name="scala-compiler" rev="2.10.6" conf="scala-tool->default">
</dependency>
    <dependency org="org.scala-lang" name="scala-compiler" rev="2.10.6" conf="scala-tool->optional(default)">
</dependency>
    <dependency org="org.scala-lang" name="scala-library" rev="2.10.6" conf="scala-tool->default">
</dependency>
    <dependency org="org.scala-lang" name="scala-library" rev="2.10.6" conf="scala-tool->optional(default)">
</dependency>
```

These are produced by 0.13.16 too, when calling `deliverLocal`

**Comment by mr-git** [comment]
tried to use that generated `ivy.xml` - it fails miserably
> java.text.ParseException: [xml parsing: ivy-\<version\>.xml.original:12:8: cvc-complex-type.2.1: Element 'conf' must have no character or element information item [children], because the type's content type is empty. in file:/Users/.../ivy-\<version\>.xml.original

**Comment by mr-git** [comment]
Will wait for 1.0.1 with clean `makeIvyXml`.

**Comment by cunei** [comment]
I investigated the matter a bit. The actual behavior of `deliver` has not changed significantly in 1.0. Most of the issues arise indeed from the change from `deliverConfiguration` to `publishConfiguration`. In particular, `publishConfiguration` is initialized using this code: https://github.com/sbt/sbt/blob/v1.0.0/main/src/main/scala/sbt/Defaults.scala#L1888-L1890
In that section, `packagedArtifacts` is invoked, which creates all the jars (that was not the case in 0.13). Also, `publishTo` is inspected; that is the reason why not defining it makes it impossible to call `deliver`. However, all that happens before `deliver` is even invoked, and neither of those values are actually needed in order to produce the `ivy.xml` file.
An easy way to restore `deliver` to the former behavior, while leaving all the rest untouched, is to rely on a slightly different `PublishConfiguration`, using a different constructor. The patch at https://github.com/cunei/sbt/commit/cc2c984b3c0ffca9057cf13b24f4dc36e56996da adds a suitable `makeIvyXmlConfiguration` key (I avoided the name `deliverConfiguration` to prevent confusion with the previous definition.) The key still refers to a `PublishConfiguration`, but one in which `packagedArtifacts` and `publishTo` are not involved. The `publishConfiguration` key is initialized separately for clarity; however it could in principle be derived directly from that key by using:
```
publishConfiguration := makeIvyXmlConfiguration
  .withArtifacts(packagedArtifacts.in(publish).value.toVector)
  .withResolverName(getPublishTo(publishTo.value).name)
```
With this change, `deliver` no longer needs a `publishTo` or generates the extra jars, while the existing keys, as defined in sbt 1.0, remain unchanged.
As a separate matter, the generated `ivy.xml` contains a reference to `scala-compiler`, and the reference to `scala-library` looks different from the one in `0.13.15`. The difference is:
```
<  <dependency org="org.scala-lang" name="scala-compiler" rev="2.10.6" conf="scala-tool->default,optional(default)"/>
<  <dependency org="org.scala-lang" name="scala-library" rev="2.10.6" conf="scala-tool->default,optional(default);compile->default(compile)"/>
---
>  <dependency org="org.scala-lang" name="scala-library" rev="2.10.6" conf="compile->default(compile)"/>
```
The differences are limited to the 'scala-tool' configuration, so in theory the change is harmless. @mr-git, when did you encounter that parsing error?

**Comment by mr-git** [comment]
When I tried to "consume" the generated and published plugin in other project.

**Comment by cunei** [comment]
@mr-git I would like to investigate the matter further. Ho do you "consume" the plugin? Do you just refer to it using `libraryDependency` in another project, or do you use some sort of postprocessing? It would greatly help us to have a self-contained example, which we can use to reproduce the failure. You can probably use `deliverLocal` to generate a sample failing ivy.xml that we can inspect.

**Comment by mr-git** [comment]
So, I have SBT plugin, published in Ivy repository and used in other project by declaring it in `plugins.sbt`
Let me see, if I can repeat the problem

**Comment by mr-git** [comment]
Ok, it will not be so easy to repeat, because we make publishing externally, not by SBT's `publish`.
Maybe, you could "fake" it by editing some plugin's ivy.xml file in local cache or on own repo...

As sample `ivy.xml` (generated for 2.10 using SBT 0.13.16):
```
<?xml version="1.0" encoding="UTF-8"?>
<ivy-module version="2.0" xmlns:e="http://ant.apache.org/ivy/extra">
	<info organisation="zzz" module="zzz" revision="1" status="release" publication="20170829223141" e:sbtVersion="0.13" e:scalaVersion="2.10">
		<description homepage="http://zzz">
		zzz
		</description>
	</info>
	<configurations>
		<conf name="compile" visibility="public" description=""/>
		<conf name="runtime" visibility="public" description="" extends="compile"/>
		<conf name="test" visibility="public" description="" extends="runtime"/>
		<conf name="provided" visibility="public" description=""/>
		<conf name="optional" visibility="public" description=""/>
		
		
		
		
		<conf name="sources" visibility="public" description=""/>
		<conf name="docs" visibility="public" description=""/>
		<conf name="pom" visibility="public" description=""/>
		
	</configurations>
	<publications>
		<artifact name="zzz" type="pom" ext="pom" conf="pom"/>
		<artifact name="zzz" type="jar" ext="jar" conf="compile"/>
		<artifact name="zzz" type="src" ext="jar" conf="sources" e:classifier="sources"/>
	</publications>
	<dependencies>
		
		<dependency org="org.scala-lang" name="scala-library" rev="2.10.6" conf="provided->default(compile)"/>
		<dependency org="org.scala-sbt" name="sbt" rev="0.13.16" conf="provided->default(compile)"/>
	</dependencies>
</ivy-module>
```

Same project, `ivy.xml` generated for Scala 2.10 using SBT 1.0.0 (using `deliverLocal`):
```
<?xml version="1.0" encoding="UTF-8"?>
<ivy-module version="2.0" xmlns:e="http://ant.apache.org/ivy/extra">
	<info organisation="zzz" module="zzz" revision="1" status="release" publication="20170829223448" e:sbtVersion="0.13" e:scalaVersion="2.10">
		<description homepage="http://zzz">
		zzz
		</description>
	</info>
	<configurations>
		<conf name="compile" visibility="public" description=""/>
		<conf name="runtime" visibility="public" description="" extends="compile"/>
		<conf name="test" visibility="public" description="" extends="runtime"/>
		<conf name="provided" visibility="public" description=""/>
		<conf name="optional" visibility="public" description=""/>
		<conf name="compile-internal" visibility="private" description="" extends="compile,optional,provided"/>
		<conf name="runtime-internal" visibility="private" description="" extends="runtime,optional"/>
		<conf name="test-internal" visibility="private" description="" extends="test,optional,provided"/>
		<conf name="plugin" visibility="private" description=""/>
		<conf name="pom" visibility="public" description=""/>
		<conf name="scala-tool" visibility="private" description=""/>
	</configurations>
	<publications>
		<artifact name="zzz" type="pom" ext="pom" conf="pom"/>
		<artifact name="zzz" type="jar" ext="jar" conf="compile"/>
		<artifact name="zzz" type="src" ext="jar" conf="compile" e:classifier="sources"/>
	</publications>
	<dependencies>
		<dependency org="org.scala-lang" name="scala-compiler" rev="2.10.6" conf="scala-tool->default,optional(default)"/>
		<dependency org="org.scala-lang" name="scala-library" rev="2.10.6" conf="scala-tool->default,optional(default);provided->default(compile)"/>
		<dependency org="org.scala-sbt" name="sbt" rev="0.13.16" conf="provided->default(compile)"/>
	</dependencies>
</ivy-module>
```

**Comment by mr-git** [comment]
I just noticed that in `deliverLocal` version, there are 2 "mostly harmless" lines (11th and 12th to be precise):
```
    <conf extends="compile,optional,provided" name="compile-internal" visibility="public" description="">
</conf>
```
And then one more pair:
```
    <conf extends="test,optional,provided" name="test-internal" visibility="public" description="">
</conf>
```

And the error I got was:
>java.text.ParseException: [xml parsing: ivy-<version>.xml.original:12:8: cvc-complex-type.2.1: Element 'conf' must have no character or element information item [children], because the type's content type is empty. in file:/Users/.../ivy-<version>.xml.original

Could it be, that problem is due to having `<conf ..></conf>` style instead of `<conf .. />` style? Could it be that leftover lines of `deliverLocal`, have "harmless" new line in `conf` node, which is not so harmless after all, as it is XML's `Text` entity?

Do you have `ivy.xml` DTD at hand?

**Comment by mr-git** [comment]
http://ant.apache.org/ivy/schemas/ivy.xsd - though I am not fluent in it

**Comment by cunei** [comment]
Hello,
When testing the xml that you pasted above for `deliverLocal`, the W3C XML Schema online validator at http://www.utilities-online.info/xsdvalidation says: "The XML is Well Formed and Valid." To double check, this is the output from `xmllint`:
```
$ xmllint --noout --schema ivy.xsd from100.xml 
from100.xml validates
```
However, the file that you pasted for `deliverLocal` does not contain the `</conf>` element on a separate line as you mention below. The attributes are also in a different order. Replacing the `compile-internal` conf line with those two separate lines, indeed I obtain the same `Cvc-complex-type.2.1` error during validation. Are you sure you are using the same `ivy.xml` that is produced by `deliverLocal`?

**Comment by mr-git** [comment]
Pardon, I forgot to mention here:
> (generated for 2.10 using SBT 0.13.16)

was generated using `deliver` on SBT 0.13.16

**Comment by cunei** [comment]
Let me clarify: in the second `ivy.xml` file you added, generated for Scala 2.10 using SBT 1.0.0 and `deliverLocal`, the `conf` lines are on a single line. Where did you encounter the `conf` lines split on two lines?

**Comment by mr-git** [comment]
Hm.. indeed.
Tried again in smaller test plugin locally - SBT 1.0.0, `^deliverLocal` - generated nice `ivy.xml` file, without those offending extra lines, but as soon as I added in `plugins.sbt`
```
addSbtPlugin("io.get-coursier" % "sbt-coursier" % "1.0.0-RC10")
```
`^deliverLocal` produced those extra lines...
Same for main plugin.
Seems that I will have to remove `sbt-coursier` as it messes up something when used with SBT 1.0.0, and will have to use either `^deliverLocal` or add the hack:
```
publishTo := Some(Resolver.file("hack",  file("hack")))
```

Will try "hack" version on my main plugin now.

Thanks!

**Comment by mr-git** [comment]
I went with `deliverLocal` for now and it works so far. Waiting for next release with clean `makeIvyXml`.

**Comment by mr-git** [comment]
Indeed: https://github.com/coursier/coursier/commit/8f988d4d087065cb8566b3239cd8edabd475962f

**Comment by cunei** [comment]
Ok, so this part is another instance of #3412 (fixed in Coursier 1.0.0-RC11, see https://github.com/coursier/coursier/issues/637#issuecomment-325053899). I will issue an sbt pull request with the patch to `deliver` then this ticket can be closed; thanks.

**Comment by mr-git** [comment]
Thank you!

**Comment by dwijnand** [comment]
Fixed with https://github.com/sbt/sbt/pull/3487 merged (right?)

**Comment by cunei** [comment]
I thought this ticket would auto-close when merging #3487 (there is the magic autoclose comment), but somehow it didn't. Yes, all is done.

**Comment by mr-git** [comment]
Will this get in SBT 1.0.2? When release is planned?

**Comment by dwijnand** [comment]
@cunei It didn't because that was merged into the 1.0.x branch, which isn't the default branch.

**Comment by dwijnand** [comment]
@mr-git Yes this will be in sbt 1.0.2, but there's no due date for that currently: https://github.com/sbt/sbt/milestone/31
